src/services/mediaDevicePreferences.ts

```diff
@@ -0,0 +1,182 @@
+/**
+ * @copyright Copyright (c) 2024 Maksim Sukharev <antreesy.web@gmail.com>
+ *
+ * @author Maksim Sukharev <antreesy.web@gmail.com>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * List all registered devices in order of their preferences
+ * Show whether device is currently unplugged or selected, if information is available
+ *
+ * @param devices list of available devices
+ * @param audioInputId id of currently selected audio input
+ * @param videoInputId id of currently selected video input
+ * @param audioInputList list of registered audio devices in order of preference
+ * @param videoInputList list of registered video devices in order of preference
+ */
+function listMediaDevices(devices: MediaDeviceInfo[], audioInputId: string, videoInputId: string, audioInputList: MediaDeviceInfo[], videoInputList: MediaDeviceInfo[]) {
+	const availableDevices = devices.map(device => device.deviceId).filter(id => id !== 'default')
+
+	const getDeviceString = (device: MediaDeviceInfo, index: number) => {
+		const isUnplugged = !availableDevices.includes(device.deviceId) ? ' (unplugged)' : ''
+		const isSelected = () => {
+			if (device.kind === 'audioinput') {
+				return device.deviceId === audioInputId ? ' (selected)' : ''
+			} else if (device.kind === 'videoinput') {
+				return device.deviceId === videoInputId ? ' (selected)' : ''
+			}
+		}
+		return `    ${index + 1}. ${device.label || device.deviceId}` + isUnplugged + isSelected()
+	}
+
+	// eslint-disable-next-line no-console
+	console.log(`Media devices:
+  Audio input:
+${audioInputList.map(getDeviceString).join('\n')}
+
+  Video input:
+${videoInputList.map(getDeviceString).join('\n')}
+`)
+}
+
+/**
+ * Modify devices list.
+ *
+ * @param device device
+ * @param devicesList list of registered devices in order of preference
+ * @param promote whether device should be promoted (to be used in updateMediaDevicesPreferences)
+ */
+function registerNewDevice(device: MediaDeviceInfo, devicesList: MediaDeviceInfo[], promote: boolean = false) {
+	const newDevicesList = devicesList.slice()
+	console.debug('Registering new device:', device)
+
+	if (promote) {
+		newDevicesList.unshift(device)
+	} else {
+		newDevicesList.push(device)
+	}
+
+	return newDevicesList
+}
```

Comment:
You should affect the original list in the argument `devicesList` as you are not reassigning the output array to the list in the function call

Reply:
just a note, preference list grows infinitely and we don't remove devices ( which is expected). Yet, we should consider edge cases and clean after a month for example (store last cleaning date) ?

Reply:
I wouldn't be too concerned about it, as we're limited by one machine